diff --git a/SpiceXPI/src/plugin/plugin.cpp b/SpiceXPI/src/plugin/plugin.cpp
index 1737fd8..be26a19 100644
--- a/SpiceXPI/src/plugin/plugin.cpp
+++ b/SpiceXPI/src/plugin/plugin.cpp
@@ -870,6 +870,8 @@ void *nsPluginInstance::ControllerWaitHelper(void *opaque)
     fake_this->CallOnDisconnected(exit_code);
     fake_this->m_external_controller.Disconnect();
     fake_this->m_pid_controller = -1;
+    unlink(fake_this->m_trust_store_file.c_str());
+    fake_this->m_trust_store_file.clear();
     return NULL;
 }
 
