From 481c3f4fbe91cb7404e76ac374d71523aa0533d7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@redhat.com>
Date: Fri, 7 Mar 2014 15:52:08 +0200
Subject: [PATCH 2/2] exit if already connected

---
 SpiceXPI/src/plugin/controller.cpp |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/SpiceXPI/src/plugin/controller.cpp b/SpiceXPI/src/plugin/controller.cpp
index 765b19e..26f9575 100644
--- a/SpiceXPI/src/plugin/controller.cpp
+++ b/SpiceXPI/src/plugin/controller.cpp
@@ -110,6 +110,8 @@ int SpiceController::Connect()
     int rc = connect(m_client_socket, (struct sockaddr *) &remote, strlen(remote.sun_path) + sizeof(remote.sun_family));
     if (rc == -1)
     {
+	if (errno == EISCONN)
+            return 1;
         QErrorHandler(errno, "connect error");
         LOG_DEBUG("Connect Error");
     }
@@ -130,6 +132,8 @@ int SpiceController::Connect(const int nRetries)
     for (int i = 0; rc != 0 && i < nRetries; ++i)
     {
         rc = Connect();
+	if (rc == 1)
+	     break;
         sleep(sleep_time);
     }
 
-- 
1.7.1

