diff --git a/SpiceXPI/src/plugin/Makefile.am b/SpiceXPI/src/plugin/Makefile.am
index 08d89d7..6ad8384 100644
--- a/SpiceXPI/src/plugin/Makefile.am
+++ b/SpiceXPI/src/plugin/Makefile.am
@@ -61,10 +61,10 @@ nsISpicec.xpt: nsISpicec.idl
 	$(AM_V_GEN)$(XPIDL) -m typelib -I $(SDK_INCLUDE_DIR) $<
 else
 nsISpicec.h: nsISpicec.idl
-	$(AM_V_GEN)$(PYTHON) `pkg-config --variable=sdkdir libxul`/sdk/bin/header.py --cachedir . -I $(SDK_INCLUDE_DIR) $< -o $@
+	$(AM_V_GEN)$(PYTHON) `pkg-config --variable=libdir libxul`/header.py --cachedir . -I $(SDK_INCLUDE_DIR) $< -o $@
 
 nsISpicec.xpt: nsISpicec.idl
-	$(AM_V_GEN)$(PYTHON) `pkg-config --variable=sdkdir libxul`/sdk/bin/typelib.py --cachedir . -I $(SDK_INCLUDE_DIR) $< -o $@
+	$(AM_V_GEN)$(PYTHON) `pkg-config --variable=libdir libxul`/typelib.py --cachedir . -I $(SDK_INCLUDE_DIR) $< -o $@
 endif
 
 BUILT_SOURCES =					\
diff --git a/configure.ac b/configure.ac
index e49e683..d251a2c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -67,8 +67,8 @@ AC_SUBST(XPIDL)
 
 if ! test -x "$XPIDL"; then
   AM_PATH_PYTHON
-  AC_CHECK_FILE([`pkg-config --variable=sdkdir libxul`/sdk/bin/header.py])
-  AC_CHECK_FILE([`pkg-config --variable=sdkdir libxul`/sdk/bin/typelib.py])
+  AC_CHECK_FILE([`pkg-config --variable=libdir libxul`/header.py])
+  AC_CHECK_FILE([`pkg-config --variable=libdir libxul`/typelib.py])
 fi
 
 XUL_INCLUDEDIR=`pkg-config --variable=includedir libxul`
